assembly/src/assembler/instruction/procedures.rs

```diff
@@ -1,131 +1,93 @@
+                            })
+                        }
+                    })?;
+                context.register_external_call(&proc, false)?;
+            }
+            Some(proc) => context.register_external_call(&proc, false)?,
+            None if matches!(kind, InvokeKind::SysCall) => {
+                return Err(AssemblyError::UnknownSysCallTarget {
+                    span,
+                    source_file: current_source_file,
+                    callee: mast_root,
+                });
+            }
+            None => context.register_phantom_call(Span::new(span, mast_root))?,
         }
 
-        // create a new CALL block for the procedure call and return
-        Ok(Some(CodeBlock::new_call(*mast_root)))
-    }
-
-    pub(super) fn call_imported(
-        &self,
-        proc_id: &ProcedureId,
-        context: &mut AssemblyContext,
-    ) -> Result<Option<CodeBlock>, AssemblyError> {
-        // make sure the procedure is in procedure cache
-        self.ensure_procedure_is_in_cache(proc_id, context)?;
-
-        // get the procedure from the assembler
-        let proc_cache = self.proc_cache.borrow();
-        let proc = proc_cache.get_by_id(proc_id).expect("procedure not in cache");
-
-        // register a "non-inlined" call to the procedure; this updates the callset of the
-        // procedure currently being compiled
-        context.register_external_call(proc, false)?;
-
-        // create a new CALL block for the procedure call and return
-        Ok(Some(CodeBlock::new_call(proc.mast_root())))
-    }
-
-    pub(super) fn syscall(
-        &self,
-        proc_id: &ProcedureId,
-        context: &mut AssemblyContext,
-    ) -> Result<Option<CodeBlock>, AssemblyError> {
-        // fetch from proc cache and check if its a kernel procedure
-        // note: the assembler is expected to have all kernel procedures properly inserted in the
-        // proc cache upon initialization, with their correct procedure ids
-        let proc_cache = self.proc_cache.borrow();
-
-        let proc = proc_cache
-            .get_by_id(proc_id)
-            .ok_or_else(|| AssemblyError::kernel_proc_not_found(proc_id))?;
-
-        // since call and syscall instructions cannot be executed inside a kernel, a callset for
-        // a kernel procedure must be empty.
-        debug_assert!(proc.callset().is_empty(), "non-empty callset for a kernel procedure");
-
-        // register a "non-inlined" call to the procedure; this updates the callset of the
-        // procedure currently being compiled
-        context.register_external_call(proc, false)?;
-
-        // create a new SYSCALL block for the procedure call and return
-        Ok(Some(CodeBlock::new_syscall(proc.mast_root())))
+        let block = match kind {
+            // For `exec`, we use a PROXY block to reflect that the root is
+            // conceptually inlined at this location
+            InvokeKind::Exec => CodeBlock::new_proxy(mast_root),
```

Comment:
Using proxy blocks is not exactly the same as inlining. The main difference is that all inlined code could go into a single `SPAN` block, while every `PROXY` block is a different node in the MAST. If the invocation target is a big procedure, the impact is probably limited, but for small procedures using a proxy block would result in significantly worse performance.

To give an example:
```
proc.foo
    mul add
end

being
    push.1.2.3.4.5.6.7.8
    exec.foo
    exec.foo
    exec.foo
    exec.foo
end
```

With the previous strategy, this would have resulted in a single span block. But now, we'll have a MAST with 7 or 8 notes.

In the transaction kernel, we actually invoke small procedures pretty frequently - so, I would expect that with this approach we'd get a much more "verbose" MAST.

This is not to say that inlining is always the right strategy. There is probably some threshold such that if a procedure is small enough, we should inline it, but if it is over this threshold, we should use a proxy block. In the future, we'll need to figure out what this threshold is and if there are any other factors which go into this decisions.

Reply:
Yes, the effect here is that procedures are not literally inlined any longer, to allow for maximal sharing of procedure MASTs. One thing this does enable now is that we can inline only when the cost is worth doing so. It's strictly more flexible/powerful (i.e. with some relatively simple heuristics, we could make automatic inlining decisions such that this approach is always more optimal than the previous approach).

One thing to note however is that by literally inlining the callee, I believe the hash of that node changes (since you'd be stripping off the outermost SPAN/JOIN/etc, and inlining the code into the SPAN of the caller). This means that code no longer is shared with other callers. That's not an issue per se, but is definitely a cost to consider with any heuristics we implement.

We have a couple of options:

1.) Implement a basic heuristic for automatic inlining based on the size of the callee procedure. We could be pretty precise about this, i.e. determining exactly where the size/performance boundary is, and choosing an option on the smaller side of the boundary. The alternative approach would be to pick a heuristic based on the procedures we would like to ensure always qualify (i.e. define it such that those procedures are always inlined by the heuristic).
2.) Add syntax for declaring procedure metadata-like attributes (we have other uses for this as well, but this is another one), then use an attribute to mark inlining preferences for procedures, e.g. `#[inline]` or `#[inline(always)]` similar to Rust. The former would indicate that we would prefer to inline, so if it is close to the boundary our heuristic uses, it will "stretch" the heuristic to include that procedure. The latter would indicate that we always want to inline regardless of the heuristic.
3.) Find a more efficient way to represent proxy nodes of this kind, i.e. actually supporting `exec` as an instruction in MAST, so that we can just lower `exec` as-is, with the semantics defined as being equivalent to a PROXY. Not sure whether this one is doable or not, but would probably be the "ideal" approach in terms of tradeoffs (it would be strictly better than doing any inlining at all).

As an aside, we could also use LLVM's heuristic for inlining functions (which IIRC is based on instruction count) as a default, and modify it as needed to better suit our environment - but I think we should be able to pin down the exact line where the choice between inlining into a SPAN or using a PROXY is the preferred choice, since that seems like the only tradeoff here. LLVM has to consider other things, such as register pressure, instruction cache, etc., that are just not things we have to deal with.

Reply:
I think the above is a perfect description for a new issue :)

> I believe the hash of that node changes (since you'd be stripping off the outermost SPAN/JOIN/etc, and inlining the code into the SPAN of the caller).

Correct. We'll need to make sure that w/e heuristic we use is fully deterministic.

Regarding the 3 options:
- Option 3 is probably a long shot - or at least I don't know how to do this yet.
- Both options 1 and 2 are good and we can support both. One question is which one to start with - and my thinking here is that we can start with w/e is simpler.

Regarding the heuristic - we'll need to think through this a bit but some things can be pretty straight-forward. A proxy node adds at least one node to the MAST and so this requires at least one extra permutation of the hash function. In the same permutation we could absorb one operation batch (up to 72 operations or up to 7 instructions with operation values) into the span block. So, if procedure body consists of a single operation batch, it should always be strictly better to inline it.

We can also argue that in most cases, even if the procedure consists of 2 or 3 operation batches it is better to inline it - but beyond that, I think things get a bit more tricky and require more analysis.

Reply:
> I think the above is a perfect description for a new issue :)

I'll make sure to create one then :). I think the underlying issue here consists of two parts: 1.) determining whether my option 3 is possible to implement; 2.) if not (or possibly in addition to), then we need to implement support for explicit and implicit inlining options (including specifying that a procedure should never be inlined, i.e. `#[inline(never)]`). The implicit choice requires an optimal heuristic, and we need to identify what that is. Does that sound about right?

> One question is which one to start with - and my thinking here is that we can start with w/e is simpler.

I think the simplest is the more explicit (i.e. manually specifying inlining preferences via attributes). It requires no additional analysis, and is naturally more flexible (i.e. we don't have to tweak the heuristic and ship a new version of the assembler every time we find an edge case, you can just note the edge case as an issue to address, and add manual annotations until the next release).

The only downside to that approach is the need to add the annotations/attributes to all procedures that we feel strongly _must_ be inlined right now. On the one hand, it's a chore, but on the other, it does provide us with an opportunity to determine what meets the threshold for "must inline", which we can then feed into our choice of heuristic.

The language support for annotations/attributes is the easy part IMO, easier than implementing the heuristic approach first anyway.

> A proxy node adds at least one node to the MAST and so this requires at least one extra permutation of the hash function. In the same permutation we could absorb one operation batch

I think there are cases where it makes sense to keep the PROXY even if the PROXYied procedure has a body that meets the inlining threshold you outlined. To better illustrate what I mean, consider the following program:

```
#! geo.masm

# [b, a, c]
export.is_pythagorean_triangle
    pow2           # b^2
    swap pow2 # a^2
    add              # a^2 + b^2
    swap pow2 # c^2
    eq                # a^2 + b^2 = c^2
end

#! main.masm

use geo

begin
    push.42.40.58
    exec.geo::is_pythagorean_triangle
    ...
end
```

Here, there is little benefit to inlining `is_pythagorean_triangle` even though its body is under the threshold for inlining you outlined. This is because it is only invoked once, and the work of computing the hash for the body of that procedure has already been done. Another permutation is required for the PROXY, but does it outweigh the cost of inlining the procedure and recomputing the hash of the body of `is_pythagorean_triangle`, _and_ the increase in code size due to duplicating its body (let's assume here that the `geo` module is a commonly used library)? That's the part I'm less confident about.

I think it probably makes sense to always inline procedures smaller than 8 ops, regardless of what I just said, except when doing so would cause the SPAN of the caller to grow too large, requiring an additional permutation anyway, at which point it's strictly all downside to inline the callee.

I'm sure we can come up with something reasonable though. We can also discuss this in more detail in the issue we create to flesh this out. I guess the question for now is, how would you like to proceed with regards to this PR? I think we should try and merge it as-is, but then follow it up with a PR to add the inlining annotations, along with the inlining pass that uses them.

Reply:
For this PR, I think we can leave things as they are. And maybe we actually create 2 issues to tackle later:

1. For adding annotations to procedures (I agree, this is probably the best thing to do first).
2. Explore heuristic for auto-inlining (or find a way not to need it at all).